crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -2107,6 +2107,248 @@ impl<'w> EntityWorldMut<'w> {
+    /// ```ignore
+    /// world.entity_mut(entity).clone_and_spawn_with(|builder| {
+    ///     builder.deny_all().allow::<ComponentB>();
+    /// });
+    /// ```
+    ///
+    /// See the methods on [`EntityCloneBuilder`] for more options.
+    ///
+    /// # Panics
+    ///
+    /// If this entity has been despawned while this `EntityWorldMut` is still alive.
+    pub fn clone_and_spawn_with(
+        &mut self,
+        f: impl FnOnce(&mut EntityCloneBuilder) + Send + Sync + 'static,
+    ) -> Entity {
+        self.assert_not_despawned();
+
+        let entity_clone = self.world.entities.reserve_entity();
+        self.world.flush();
+
+        let mut builder = EntityCloneBuilder::new(self.world);
+        f(&mut builder);
+        builder.clone_entity(self.entity, entity_clone);
+
+        self.world.flush();
+        self.update_location();
+        entity_clone
+    }
+
+    /// Clones the specified components of this entity and inserts them into another entity.
+    ///
+    /// Components can only be cloned if they implement
+    /// [`Clone`] or [`Reflect`](bevy_reflect::Reflect).
+    ///
+    /// # Panics
+    ///
+    /// - If this entity has been despawned while this `EntityWorldMut` is still alive.
+    /// - If the target entity does not exist.
+    pub fn clone_components<B: Bundle>(&mut self, target: Entity) -> &mut Self {
+        self.assert_not_despawned();
+
+        let mut builder = EntityCloneBuilder::new(self.world);
+        builder.deny_all().allow::<B>();
+        builder.clone_entity(self.entity, target);
+
+        self.world.flush();
+        self.update_location();
+        self
+    }
+
+    /// Clones the specified components of this entity with those components' required components
+    /// and inserts them into another entity.
+    ///
+    /// Components can only be cloned if they implement
+    /// [`Clone`] or [`Reflect`](bevy_reflect::Reflect).
+    ///
+    /// # Panics
+    ///
+    /// - If this entity has been despawned while this `EntityWorldMut` is still alive.
+    /// - If the target entity does not exist.
+    pub fn clone_components_with_requires<B: Bundle>(&mut self, target: Entity) -> &mut Self {
+        self.assert_not_despawned();
+
+        let storages = &mut self.world.storages;
+        let components = &mut self.world.components;
+        let bundles = &mut self.world.bundles;
+
+        let bundle_id = bundles.register_contributed_bundle_info::<B>(components, storages);
+        // SAFETY: the `BundleInfo` for this `BundleId` is initialized above
+        let bundle_info = unsafe { bundles.get_unchecked(bundle_id) };
+        let component_ids = bundle_info.contributed_components().to_owned();
+
+        let mut builder = EntityCloneBuilder::new(self.world);
+        builder.deny_all().allow_by_ids(component_ids);
+        builder.clone_entity(self.entity, target);
+
+        self.world.flush();
+        self.update_location();
+        self
+    }
```

Comment:
I'm pretty sure allowing and denying components with their required components can be done as a mode for `EntityCloneBuilder`, which should be more flexible:
```rs
builder.with_required_components(|builder| {
  builder
    .allow::<B>()       // Calling with_required_components can set an internal flag which would
    .allow_by_ids(ids); // make all allow* and deny* methods also include contributed_components
});
```

Reply:
Something like this?
<details open>
<summary>Code</summary>

```
pub fn with_required_components(
    &mut self,
    f: impl FnOnce(&mut EntityCloneBuilder) + Send + Sync + 'static,
) -> &mut Self {
    self.attach_required_components = true;
    f(self);
    self.attach_required_components = false;
    self
}

pub fn allow<T: Bundle>(&mut self) -> &mut Self {
    if self.filter_allows_components {
        T::get_component_ids(self.world.components(), &mut |id| {
            if let Some(id) = id {
                self.filter.insert(id);
                if self.attach_required_components {
                    if let Some(info) = self.world.components().get_info(id) {
                        for required in info.required_components().iter_ids() {
                            self.filter.insert(required);
                        }
                    }
                }
            }
        });
    } else {
        T::get_component_ids(self.world.components(), &mut |id| {
            if let Some(id) = id {
                self.filter.remove(&id);
                if self.attach_required_components {
                    if let Some(info) = self.world.components().get_info(id) {
                        for required in info.required_components().iter_ids() {
                            self.filter.remove(&required);
                        }
                    }
                }
            }
        });
    }
    self
}
```
</details>

Reply:
I think this code can actually be deduplicated a bit if we replace `if self.filter_allows_components` with something like this:
```rs
fn apply_filter(
    &mut self,
    id: ComponentId,
) {
    if self.filter_allows_components {
        self.filter.insert(id);
    } else {
        self.filter.remove(&id);
    }
}
```

Reply:
I can't get around the borrow checker to get that to work, but moving the `if`s around is still an improvement